internal/controller/apiextensions/composite/composition_render.go

```diff
@@ -80,26 +80,19 @@ func RenderFromJSON(o resource.Object, data []byte) error {
 	return nil
 }
 
-// RenderFromCompositePatches renders the supplied composed resource by applying
-// all patches that are _from_ the supplied composite resource.
-func RenderFromCompositePatches(cd resource.Composed, xr resource.Composite, p []v1.Patch) error {
+// RenderFromCompositeAndEnvironmentPatches renders the supplied composed
+// resource by applying all patches that are _from_ the supplied composite
+// resource or are from or to the supplied environment.
+func RenderFromCompositeAndEnvironmentPatches(cd resource.Composed, xr resource.Composite, e *Environment, p []v1.Patch) error {
 	for i := range p {
 		if err := Apply(p[i], xr, cd, patchTypesFromXR()...); err != nil {
 			return errors.Wrapf(err, errFmtPatch, p[i].Type, i)
 		}
-	}
-	return nil
-}
 
-// RenderToAndFromEnvironmentPatches renders the supplied composed resource by
-// applying all patches that are from or to the supplied environment.
-func RenderToAndFromEnvironmentPatches(cd resource.Composed, e *Environment, p []v1.Patch) error {
-	if e == nil {
-		return nil
-	}
-	for i := range p {
-		if err := ApplyToObjects(p[i], e, cd, patchTypesFromToEnvironment()...); err != nil {
-			return errors.Wrapf(err, errFmtPatch, p[i].Type, i)
+		if e != nil {
```

Comment:
I'm probably missing something, but if the from composite patches are applied first, won't the from environment patches override the composite values in cases where they overlap?  I thought we wanted the environment patches to serve as a default that can be overridden by the claim/composite?  Like I said I'm probably missing something.

Reply:
That's how it works in v1.14. In earlier versions the environment patches were executed in order with the regular from-composite patches.

This PR would restore the previous behaviour.

Reply:
Bob, i think the fix here is to process(apply) the patches `p` in a single pass (loop) rather than the 2 separate loops (1.14 change, which results in the overridden behaviour). the `only` filter last arg to Apply/ApplyToObjects methods ensures that a patch can be either a FromXR types patch or a Environment related patch types.

Reply:
@ravilr thanks - that's what I was missing.